node/src/reactor/main_reactor/tests/binary_port.rs

```diff
@@ -0,0 +1,669 @@
+    asserter: Box<dyn Fn(&BinaryResponse) -> bool>,
+}
+
+fn validate_metadata(
+    response: &BinaryResponse,
+    expected_payload_type: Option<PayloadType>,
+) -> bool {
+    response.is_success()
+        && response.returned_data_type_tag()
+            == expected_payload_type.map(|payload_type| payload_type as u8)
+        && expected_payload_type.map_or(true, |_| !response.payload().is_empty())
+}
+
+fn validate_deserialization<T>(response: &BinaryResponse) -> Option<T>
+where
+    T: FromBytes,
+{
+    FromBytes::from_bytes(response.payload())
+        .ok()
+        .map(|(data, remainder)| {
+            assert!(remainder.is_empty());
+            data
+        })
+}
+
+fn assert_response<T, F>(
+    response: &BinaryResponse,
+    payload_type: Option<PayloadType>,
+    validator: F,
+) -> bool
+where
+    T: FromBytes,
+    F: FnOnce(T) -> bool,
+{
+    validate_metadata(response, payload_type)
+        && payload_type.map_or(true, |_| {
+            validate_deserialization::<T>(response).map_or(false, validator)
+        })
+}
+
+#[tokio::test]
+async fn binary_port_component() {
+    testing::init_logging();
+
+    let (
+        client,
+        (finish_cranking, mut rng, network_chainspec_raw_bytes, highest_block, secret_key),
+    ) = setup().await;
+
+    let test_cases = &[
+        block_hash_2_height(),
+        highest_complete_block(),
+        completed_blocks_contain_true(),
+        completed_blocks_contain_false(),
+        transaction_hash_2_block_hash_and_height(&mut rng),
+        peers(),
+        uptime(),
+        last_progress(),
+        reactor_state(),
+        network_name(),
+        consensus_validator_changes(),
+        block_synchronizer_status(),
+        available_block_range(),
+        next_upgrade(),
+        consensus_status(),
+        chainspec_raw_bytes(network_chainspec_raw_bytes),
+        node_status(),
+        get_block_header(highest_block.block().clone_header()),
+        get_block_transfers(highest_block.block().clone_header()),
+        get_era_summary(*highest_block.block().state_root_hash()),
+        get_all_bids(*highest_block.block().state_root_hash()),
+        get_trie(*highest_block.block().state_root_hash()),
+        try_accept_transaction(&secret_key),
+        try_accept_transaction_invalid(&mut rng),
+        try_spec_exec_invalid(&mut rng, highest_block.block().clone_header()),
+    ];
+
+    let header = BinaryRequestHeader::new(ProtocolVersion::V1_0_0);
+    let header_bytes = ToBytes::to_bytes(&header).expect("should serialize");
+
+    for TestCase {
+        name,
+        request,
+        asserter,
+    } in test_cases
+    {
+        let original_request_bytes = header_bytes
+            .iter()
+            .chain(
+                ToBytes::to_bytes(&request)
+                    .expect("should serialize")
+                    .iter(),
+            )
+            .cloned()
+            .collect::<Vec<_>>();
+
+        let request_guard = client
```

Comment:
> Also, it might have been nice to have a separate test instance per test case so we benefit from parallelisation in the runner.

I think that in this case making a separate test for every use case will make the whole process slower. The scenario is:
1. setup network (_fast_)
2. wait until height 2 (_slow_)
3. send a bunch of requests to binary port (_fast_)

If we split this into separate tests, we'd need to repeat points 1 and 2 for every case.

Reply:
> I think that in this case making a separate test for every use case will make the whole process slower.

Good point - agreed.

However, I still think we should have timeouts for the tests to avoid hanging.